### PR TITLE
Don't duplicate resources that are referenced multiple times

### PR DESCRIPTION
**Pull request by lilleyse**
Fixes https://github.com/AnalyticalGraphicsInc/gltf-pipeline/issues/416

Follow up of https://github.com/AnalyticalGraphicsInc/gltf-pipeline/pull/474 which was closed. This has the same basic idea but extends it to work with separate resources.

### CODE REVIEW

**Comment by ViliusLuneckas** [comment]
Hi @lilleyse, any plans to merge it soon and release to `npm` registry?

File: lib/writeResources.js [hunk 1]
```diff
@@ -112,13 +115,26 @@ function writeDataUri(object, extension) {
     object.uri = 'data:' + mimeType + ';base64,' + source.toString('base64');
 }
 
-function writeBufferView(gltf, object) {
+function writeBufferView(gltf, object, writtenResourceMap) {
     delete object.uri;
+
+    // If we've written this resource before, re-use the bufferView
+    const resourceId = object.extras._pipeline.resourceId;
+    if (defined(resourceId) && defined(writtenResourceMap[resourceId])) {
```
Comment (likangning93) [inline lib/writeResources.js]: From `readResources.js` it looks like `resourceId` isn't supposed to be defined here if `writtenResourceMap` isn't refined, but should this also check `defined(writtenResourceMap)`?

File: lib/writeResources.js [hunk 2]
```diff
@@ -183,12 +199,25 @@ function getRelativePath(gltf, object, index, extension, options) {
     return relativePath;
 }
 
-function writeFile(gltf, object, index, extension, options) {
+function writeFile(gltf, object, index, extension, writtenResourceMap, options) {
     delete object.bufferView;
+
+    // If we've written this resource before, re-use the uri
+    const resourceId = object.extras._pipeline.resourceId;
+    if (defined(resourceId) && defined(writtenResourceMap[resourceId])) {
```
Comment (likangning93) [inline lib/writeResources.js]: Similar to above, should this check `defined(writtenResourceMap)`?

**Comment by lilleyse** [comment]
@likangning93 all images, shaders, buffers now pass through `writtenResourceMap`, but yeah, ultimately only images and shaders will use it, not buffers.

This change handles all your comments and makes https://github.com/CesiumGS/gltf-pipeline/pull/483#discussion_r399859575 OBE.

File: lib/writeResources.js [hunk 3]
```diff
@@ -46,32 +46,35 @@ function writeResources(gltf, options) {
...
     });
 
     ForEach.shader(gltf, function(shader, i) {
-        writeShader(gltf, shader, i, options);
+        writeShader(gltf, shader, i, writtenResourceMap, options);
     });
 
     // Buffers need to be written last because images and shaders may write to new buffers
-    removeUnusedElements(gltf, ['acessor', 'bufferView', 'buffer']);
+    removeUnusedElements(gltf, ['accessor', 'bufferView', 'buffer']);
```
Comment (lilleyse) [inline lib/writeResources.js]: 🤦‍♂ 

Got a vscode spell checker plugin which came to the rescue here

**Comment by likangning93** [comment]
Thanks @lilleyse!

**Pull request merged**